Add Containerfile for building from osbuild and images sources for development

achilleas-k opened the pull request:
New Containerfile that requires build contexts pointing to checked out sources for osbuild/osbuild and osbuild/images to make building development versions of all the components more convenient.

PR includes a README describing how to build the container.

bcl commented on devel/README.md:
````diff
@@ -0,0 +1,20 @@
+```
+the container can be built using:
+```
+$ podman build . --build-context=osbuild=$HOME/src/osbuild --build-context=images=$HOME/src/images --build-context=bib=.. -t bootc-image-builder:devel
````
It feels more natural to me to run this from the root of bib, so instead of the extra build-context for bib use `-f ./devel/Containerfile` and drop the `--from=bib` from it.

achilleas-k replied:
Hm yeah, I like that more too.  I'll update.

achilleas-k replied:
Updated Containerfile and instructions.

cgwalters reviewed (commented):
Gave this a try and hit:

```
[2/4] STEP 5/5: RUN make rpm
fatal: detected dubious ownership in repository at '/build'
To add an exception for this directory, call:
```

We've got a workarounds for this in various CI systems that I've touched; I think it happens here because the container build is run as uid 0 but my git repos are owned on the host by a nonzero uid.

Does it work for you?  What's your setup?

achilleas-k commented:
> I think it happens here because the container build is run as uid 0 but my git repos are owned on the host by a nonzero uid.

Oh that old nugget.

Yes, git doesn't like running under one uid when the repository is owned by another.  Fixed.

cgwalters commented:
So far this seems to be working for me!  One notable papercut (unrelated to this) is that I'm used to working on Go projects that vendor, and downloading go modules (especially since we're not setting `GOPROXY=https://proxy.golang.org` for example) is just very slow.  I think we could probably adjust this to add a cache volume ([like this](https://docs.docker.com/engine/reference/builder/#example-cache-go-packages)?  haven't tested)

achilleas-k commented:
I'd be more in favour of vendoring the dependencies, which should speed up builds both here and in the standard case without needing to do the whole `go.mod` juggling.
@ondrejbudai wdyt?  Is it time?

achilleas-k commented:
> I think we could probably adjust this to add a cache volume ([like this](https://docs.docker.com/engine/reference/builder/#example-cache-go-packages)? haven't tested)

I tested this with:
```diff
diff --git a/devel/Containerfile b/devel/Containerfile
index cab6385..e9db5b6 100644
--- a/devel/Containerfile
+++ b/devel/Containerfile
@@ -22,7 +22,7 @@ WORKDIR /build/bib
 RUN go mod edit -replace github.com/osbuild/images=../images
 RUN go mod tidy
 WORKDIR /build
-RUN ./build.sh
+RUN --mount=type=cache,target=/root/.cache/go-build ./build.sh
 
 
 FROM registry.fedoraproject.org/fedora:39
```
and I'm not sure whether it's helping.  I made sure the go-cache for root was fresh by building `osbuid/images/cmd/build` and `osbuild/bootc-image-builder/cmd/bootc-image-builder` as root and then built the container, but it doesn't seem to be helping.  Did you have any better luck?

achilleas-k commented:
@cgwalters I ran into the same issue you did and was worried we broke something.  I ended up building the container with the same versions as the one in the container that works and that was also failing.  In the end the issue seems to occur when `/store` isn't declared as a `VOLUME` in the `Containerfile`.
Updated `devel/Containerfile` now and it should work.

FTR:
```
› git diff
diff --git a/devel/Containerfile b/devel/Containerfile
index e36388d..eac2348 100644
--- a/devel/Containerfile
+++ b/devel/Containerfile
@@ -31,5 +31,5 @@ COPY entrypoint.sh /entrypoint.sh

 ENTRYPOINT ["/entrypoint.sh"]
 VOLUME /output
-VOLUME /store
+# VOLUME /store
 VOLUME /rpmmd

› git -C ../images rev-parse HEAD
a7122313c58186a2ba1fc7b2cc6483a0c7c8e66d

› git -C ../osbuild rev-parse HEAD
161c19601a1a76e96c56aa3dfff6e153c73ba2e0

› sudo podman build --file=devel/Containerfile --build-context=osbuild=../osbuild --build-context=images=../images -t bootc-image-builder:devel .
...
Successfully tagged localhost/bootc-image-builder:devel
7c29a68a5bb8c13adfb3a67f420b155fda02dd93c572b3cd3c132d1a75d0e404


› sudo podman run \
    --rm \
    -it \
    --privileged \
    --pull=newer \
    --security-opt label=type:unconfined_t \
    --volume ./output:/output \
    bootc-image-builder:devel \
    --type qcow2 \
    quay.io/centos-bootc/centos-bootc:stream9
Generating manifest-qcow2.json ... DONE
Building manifest-qcow2.json
...
Pipeline build: 793d878eaa9b5f942e60299b422365f30285b80b0bab8bf43cbc8626bbafbbbc
Build
  root: <host>
  runner: org.osbuild.fedora38 (org.osbuild.fedora38)
org.osbuild.container-deploy: ed4e88e7df0fa547c37fe797447b930d2a5dcfcf6722ca84488e9b60c4ac3503 {
  "exclude": [
    "/sysroot"
  ]
}
...
Copying config sha256:4b40b903d3fefb478294678a59fb4860b141ccd90b8087e03ca5866667e830e9
Writing manifest to image destination
Untagged: docker.io/library/tmp-container-deploy-09606772804221:latest
Deleted: 4b40b903d3fefb478294678a59fb4860b141ccd90b8087e03ca5866667e830e9
Traceback (most recent call last):
  File "/run/osbuild/bin/org.osbuild.container-deploy", line 96, in <module>
    r = main(args["inputs"], args["tree"], args["options"])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/run/osbuild/bin/org.osbuild.container-deploy", line 87, in main
    with mount_container(image_tag) as img:
  File "/usr/lib64/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/run/osbuild/bin/org.osbuild.container-deploy", line 58, in mount_container
    raise RuntimeError(f"Failed to mount image ({code}): {msg}")
RuntimeError: Failed to mount image (125): time="2024-02-07T15:58:51Z" level=error msg="Unmounting /var/lib/containers/storage/overlay/920a86e18e78590d8d33d0c399fb786ebecf425ced25e9da923b54b285d0b5bf/merged: invalid argument"
Error: creating overlay mount to /var/lib/containers/storage/overlay/920a86e18e78590d8d33d0c399fb786ebecf425ced25e9da923b54b285d0b5bf/merged, mount_data="lowerdir=/var/lib/containers/storage/overlay/920a86e18e78590d8d33d0c399fb786ebecf425ced25e9da923b54b285d0b5bf/diff:/var/lib/containers/storage/overlay/l/67OQCYKK6SKXVZUKB2Z5INBQHF:/var/lib/containers/storage/overlay/l/S3EIY2IRL7ZVSGVXXYYLE2ELHQ:/var/lib/containers/storage/overlay/l/CGYGO2TU4VJ3QJ4MIB7XV7CDOU:/var/lib/containers/storage/overlay/l/DU2HKU5P3T5FX4VOACYFMIF5ZQ:/var/lib/containers/storage/overlay/l/5NK2BEQPWIUDMZFPFC5JWQJJLS:/var/lib/containers/storage/overlay/l/WFC7EY3CUPTR5A6UXG2XRRXUSV:/var/lib/containers/storage/overlay/l/TQJLAONUICKHUYDGE3TFZTWPJH:/var/lib/containers/storage/overlay/l/X2F5N7WDYPUEVOVSK44Y5YB6UW:/var/lib/containers/storage/overlay/l/HB75UDYSMRXG3HLBEPKICCPRU5:/var/lib/containers/storage/overlay/l/HIXAF4D3RIVWL6E4BMZMJERYBU:/var/lib/containers/storage/overlay/l/OLUX4KX2PAYF4MTQJS6Q53S2EF:/var/lib/containers/storage/overlay/l/XWBKKZ6N3DTALOLXWD3FKWOIYR:/var/lib/containers/storage/overlay/l/54RPU4Q2BHCVPNDEF6HGTW3DXP:/var/lib/containers/storage/overlay/l/FCZREVJHQLXJP2GP4MJ2WFY7X3:/var/lib/containers/storage/overlay/l/ENCI5TH4WIQO2XMW2FSOIHOVZE:/var/lib/containers/storage/overlay/l/POQTIA5GSIPOWNNCUGFOJJ2IQS:/var/lib/containers/storage/overlay/l/ZHHCUMHPBP32KODKSXBQZELXSP:/var/lib/containers/storage/overlay/l/GIQUJ4AERN7ZJM25255SC2NFNS:/var/lib/containers/storage/overlay/l/UYQUAHJXG2EVTRNDYNS2RQGD7X:/var/lib/containers/storage/overlay/l/QTXASMENXRK73OG45WA7Z3NJUK:/var/lib/containers/storage/overlay/l/GJZUICOZGQVE3LOCNRB5SJT3EI:/var/lib/containers/storage/overlay/l/MW7RBIZ7APFDQYM3BX32BNXPVS:/var/lib/containers/storage/overlay/l/VZZZ64EUNGNMF3YLLMSUYS42NP:/var/lib/containers/storage/overlay/l/FZEVQVAWOCCSESET4O3NVFETF5:/var/lib/containers/storage/overlay/l/JIUTBLG67IH3IQPLVW2THPC2VT:/var/lib/containers/storage/overlay/l/LKHM2A2L6ETIADMOOHBGTP2ZNZ:/var/lib/containers/storage/overlay/l/DWGLARFEJ4E42VJHBQECCXIBN6:/var/lib/containers/storage/overlay/l/GC6L5EX4TDI6KBZ3YUZ64DJNDE:/var/lib/containers/storage/overlay/l/H3MOYBYN34XLPVLJKVLZGXFUN7:/var/lib/containers/storage/overlay/l/LDYC35DMVXBEJY5KKMRJA7QYYP:/var/lib/containers/storage/overlay/l/BDT3GLW34LVCWMADTU3U7PTTOD:/var/lib/containers/storage/overlay/l/CKJCNB33MEVOQTBPZG74YZI5J2:/var/lib/containers/storage/overlay/l/2L56RSN4DP6P76PKU5ZHCCQB3C:/var/lib/containers/storage/overlay/l/5QHGRDQ3W3TMPDTVBOCLDYU6J5:/var/lib/containers/storage/overlay/l/PGEPMOGJNJ6HFYNFDW2TCRG2CB:/var/lib/containers/storage/overlay/l/UKRWVXYFKGHTH25CDF2HZDEOXW:/var/lib/containers/storage/overlay/l/EZQOUPFZFGGQBFPEIVEKW2JRAK:/var/lib/containers/storage/overlay/l/HRCMYWGFZCDY6Z4HYLF5IGONDW:/var/lib/containers/storage/overlay/l/3SLBRONXXU7WRAFCP5JEKK2RBR:/var/lib/containers/storage/overlay/l/WTK7S4QHIEMOPAHU7YELPMK2KI:/var/lib/containers/storage/overlay/l/4H6QONCA2WA5IMS2TZF2KI6VMQ:/var/lib/containers/storage/overlay/l/6FUSEINQ2UMKWBZJAANKMVYDPK:/var/lib/containers/storage/overlay/l/FKMD44QRW3WUBKTAM4RJHLU3QP:/var/lib/containers/storage/overlay/l/KWE24OEYP3HVOXIRQMPSPJFXCV:/var/lib/containers/storage/overlay/l/BSQAI463U3JT5XDNM7P4X2EY3T:/var/lib/containers/storage/overlay/l/RFLDFS5ARSENZYU3ZQVCOGBI42:/var/lib/containers/storage/overlay/l/CXKQ3ISMSSSVWTBFUA6W76ZGUS:/var/lib/containers/storage/overlay/l/6V354YRWC34CF3KFGOL5ZO22J5:/var/lib/containers/storage/overlay/l/QYHRTTJ2D5UELUXDMJTB7NLOB5:/var/lib/containers/storage/overlay/l/6YMB6CR4MKWLJJQNLXFNKXDJW2:/var/lib/containers/storage/overlay/l/XUWJOV5CJCVFL6PX7NVBTZ5RQM:/var/lib/containers/storage/overlay/l/YQ2W7CV5PXRVMAJGQP2FK75J57:/var/lib/containers/storage/overlay/l/QA52NZSPIQSWC3LZVSJUVUKPYV:/var/lib/containers/storage/overlay/l/F5X6T6DZ6A7HVBTJZKQKEUAPCT:/var/lib/containers/storage/overlay/l/I4GLLVJN4GK7572TVCO7BNFJAH:/var/lib/containers/storage/overlay/l/R2RMNYXDUVCAUQIXMA7AM2Y3T6:/var/lib/containers/storage/overlay/l/FFEDTGRPUBCQSGKDX6WNOAOXSO:/var/lib/containers/storage/overlay/l/2E63BCH7XYHWKNOYBNAMW5TD6J:/var/lib/containers/storage/overlay/l/7ZAOA5BHE3HKFESKJKJPO77IVK:/var/lib/containers/storage/overlay/l/YM7ZSVEXCD46WGK76PMJI2LYVU:/var/lib/containers/storage/overlay/l/EH4UX276ONEYB3H2KRN5YFC47Y:/var/lib/containers/storage/overlay/l/NLBQMMWJHMMUDKQIU66ILNBGRU:/var/lib/containers/storage/overlay/l/XUYOHHIABR7E5WT2ZDM5EBQLFK:/var/lib/containers/storage/overlay/l/EBP6XDZDKRVF5WU2XRUZJLHO4T:/var/lib/containers/storage/overlay/l/CON3TF4MLZTVMTEGQE7T4A6OAU,ro": using mount program /usr/bin/fuse-overlayfs: unknown argument ignored: lazytime
fuse: device not found, try 'modprobe fuse' first
fuse-overlayfs: cannot mount: No such file or directory
: exit status 1

⏱  Duration: 25s

Failed
Error: running osbuild failed: exit status 1
2024/02/07 15:58:55 error: running osbuild failed: exit status 1
a
```

cgwalters commented:
Ahh of course, that explains it.  Nice find.

cgwalters commented:
This one intersects with https://github.com/osbuild/bootc-image-builder/pull/176 - if this merges first then I'll rework that one to also centralize the list of dependent packages in this containerfile too.
